File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ResultValueString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ResultValueString {
    private static final String DNCString = "?DNC?";

    public String sum;

    public ResultValueString(String sum) {
        this.sum = sum;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static ResultValueString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "ResultValue");
        return new ResultValueString(parts.get(0));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ResultValueString)) return false;
        ResultValueString that = (ResultValueString) o;
        return (DNCString.equals(sum) || DNCString.equals(that.sum) || Objects.equals(sum, that.sum));
    }

    @Override
    public int hashCode() {
        return Objects.hash(sum);
    }

    @Override
    public String toString() {
        return Tokens.token(sum);
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ResultValueString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class ResultValueString
    {
        public string sum;

        public ResultValueString(string sum)
        {
            this.sum = sum;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static ResultValueString FromText(string text)
        {
            var parts = Tokens.Require(text, 1, "ResultValue");
            return new ResultValueString(parts[0]);
        }

        public ResultValueTyped ToResultValueTyped()
        {
            return new ResultValueTyped(
                int.Parse(this.sum)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(sum);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not ResultValueString other) return false;
            return (DNCString == this.sum || DNCString == other.sum || object.Equals(this.sum, other.sum));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.sum);
            return h.ToHashCode();
        }
    }
}
